File: trunk/product/src/control/DC0019.php
<?php
/* N21 eBusiness Suite
 * Copyright: Nan21 Electronics srl
 * Generated content.
 * DC0019 Controller: Tasks
 */

require_once(PATH_CTRL_FRMWK."/Controller.php");

class DC0019 extends Controller {


private function preQuery(&$params, &$where) {
    if (!empty($_REQUEST["QRY_CLOSED"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "CLOSED like :CLOSED";
      $params["CLOSED"] = $_REQUEST["QRY_CLOSED"];
    }
    if (!empty($_REQUEST["QRY_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "ID like :ID";
      $params["ID"] = $_REQUEST["QRY_ID"];
    }
    if (!empty($_REQUEST["QRY_TITLE"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "TITLE like :TITLE";
      $params["TITLE"] = $_REQUEST["QRY_TITLE"];
    }
    if (!empty($_REQUEST["QRY_STATUS"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "STATUS like :STATUS";
      $params["STATUS"] = $_REQUEST["QRY_STATUS"];
    }
}

public function doQuery() {
  try {
    $start = nvl($this->getRequestParam("start"), 0);
    $limit = nvl($this->getRequestParam("limit"),20);
    $orderBy = (!empty($_REQUEST["sort"]))?$_REQUEST["sort"]:"";
    $orderSense = (!empty($_REQUEST["dir"]))?$_REQUEST["dir"]:"";
    $orderByClause = (!empty($orderBy))? "order by $orderBy $orderSense " : "" ;
    $where = "";
    $params = array();
    $this->preQuery($params, $where);
    if (!empty($where)) {
      $where = " where ".$where;
    }
    $sql = "select 
                CLOSED
                ,ID
                ,TITLE
                ,FINISH_DATE
                ,START_DATE
                ,STATUS
                ,ASSIGNED_TO
                ,CREATEDON
                ,CREATEDBY
                ,MODIFIEDON
                ,MODIFIEDBY
                ,NOTES
            from TASKS  $where $orderByClause ";
    $rs = $this->db->SelectLimit($sql, $limit, $start, $params);
    $rsCount = $this->db->Execute("select count(*) TOTALCOUNT from (".$sql.") t", $params);
    $rsCount->MoveFirst();
    $columns = array(
      "CLOSED"
      ,"ID"
      ,"TITLE"
      ,"FINISH_DATE"
      ,"START_DATE"
      ,"STATUS"
      ,"ASSIGNED_TO"
      ,"CREATEDON"
      ,"CREATEDBY"
      ,"MODIFIEDON"
      ,"MODIFIEDBY"
      ,"NOTES"
      );
    $dataOut = $this->serializeCursor($rs,$columns, $this->query_data_format);
    if ($this->query_data_format == "xml" ) {header("Content-type: application/xml");}
    print $this->formatQueryResponseData($dataOut, $rsCount->fields["TOTALCOUNT"]);
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function doQuery  */


public function doExport() {
  try {
    $start = nvl($this->getRequestParam("start"), 0);
    $limit = nvl($this->getRequestParam("limit"),20);
    $groupBy = (!empty($_REQUEST["groupBy"]))?$_REQUEST["groupBy"]:"";
    $orderBy = (!empty($_REQUEST["sort"]))?$_REQUEST["sort"]:"";
    $orderSense = (!empty($_REQUEST["dir"]))?$_REQUEST["dir"]:"";
    $orderByClause = (!empty($orderBy))? "order by $orderBy $orderSense " : "" ;
    $where = "";
    $params = array();
    $this->preQuery($params, $where);
    if (!empty($where)) {
      $where = " where ".$where;
    }
    $sql = "select 
                TITLE
                ,ASSIGNED_TO
                ,START_DATE
                ,FINISH_DATE
                ,STATUS
                ,CLOSED
                ,NOTES
                ,ID
                ,CREATEDON
                ,CREATEDBY
                ,MODIFIEDON
                ,MODIFIEDBY
            from TASKS  $where $orderByClause ";
    $rs = $this->db->Execute($sql, $params);
    $rsCount = $this->db->Execute("select count(*) TOTALCOUNT from (".$sql.") t", $params);
    $rsCount->MoveFirst();
    $columns = array(
     "TITLE"
     ,"ASSIGNED_TO"
     ,"START_DATE"
     ,"FINISH_DATE"
     ,"STATUS"
     ,"CLOSED"
     ,"NOTES"
     ,"ID"
     ,"CREATEDON"
     ,"CREATEDBY"
     ,"MODIFIEDON"
     ,"MODIFIEDBY"
      );
    if (!empty($_REQUEST["_p_disp_cols"])) {
      $columns = explode("|",$_REQUEST["_p_disp_cols"]);
    }
    $dataOut = $this->serializeCursor($rs,$columns,"xml");
    $dataOut = "<records>".$dataOut."</records>";
    $dataOut = "<queryParams>".$this->serializeArray($params,"xml")."</queryParams>".$dataOut;
    $dataOut = "<columnDef>".$this->columnDefForExport($columns,$this->fieldDef,true).$this->columnDefForExport(array_diff(array_keys($params), $columns),$this->fieldDef,false)."</columnDef>".$dataOut;
    $dataOut = "<staticText>".$this->exportLocalizedStaticText()."</staticText>".$dataOut;
    $dataOut = "<groupBy>".$groupBy."</groupBy>".$dataOut;
    $dataOut = "<reportData  title=\"".$this->getDcTitle()."\" by=\"".$_SESSION["user"]["userName"]."\" on=\"".date(DATE_FORMAT)."\">".$dataOut."</reportData>";
    $this->beginExport();
    print $dataOut;
    $this->endExport();
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function doExport  */


public function fetchRecord() {
  try {
    $RECORD = array();
    $RECORD["ID"] = $this->getRequestParam("ID");
    if (empty($RECORD["ID"])) { throw new Exception("Missing value for primary key field ID in DC0019.fetchRecord().");}
    $pkArray = array("ID" => $RECORD["ID"]);
    $this->findByPk($pkArray, $RECORD);
    print "{success:true, data:".json_encode($RECORD)."}";
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function fetchRecord */


public function doInsert() {
  $this->logger->debug("start: ".$this->dcName.".doInsert");
  try {
    $RECORD = array();
    $RECORD["ASSIGNED_TO"] = $this->getRequestParam("ASSIGNED_TO");
    $RECORD["CLOSED"] = $this->getRequestParamBoolean("CLOSED");
    $RECORD["CREATEDBY"] = $this->getRequestParam("CREATEDBY");
    $RECORD["CREATEDON"] = $this->getRequestParam("CREATEDON");
    $RECORD["FINISH_DATE"] = $this->getRequestParam("FINISH_DATE");
    $RECORD["ID"] = $this->getRequestParam("ID");
    $RECORD["MODIFIEDBY"] = $this->getRequestParam("MODIFIEDBY");
    $RECORD["MODIFIEDON"] = $this->getRequestParam("MODIFIEDON");
    $RECORD["NOTES"] = $this->getRequestParam("NOTES");
    $RECORD["START_DATE"] = $this->getRequestParam("START_DATE");
    $RECORD["STATUS"] = $this->getRequestParam("STATUS");
    $RECORD["TITLE"] = $this->getRequestParam("TITLE");
    $sql = "insert into TASKS(
                 CLOSED
                ,ID
                ,TITLE
                ,FINISH_DATE
                ,START_DATE
                ,STATUS
                ,ASSIGNED_TO
                ,CREATEDBY
                ,MODIFIEDBY
                ,NOTES
            ) values ( 
                 :CLOSED
                ,:ID
                ,:TITLE
                ,:FINISH_DATE
                ,:START_DATE
                ,:STATUS
                ,:ASSIGNED_TO
                ,:CREATEDBY
                ,:MODIFIEDBY
                ,:NOTES
    )";
    $stmt = $this->db->prepare($sql);
    $_seq = $this->db->execute("select seq_task_id.nextval seq_val from dual")->fetchRow();
    $RECORD["ID"] = $_seq["SEQ_VAL"];
    $this->logger->debug("insert of RECORD: ".$this->logger->map2string($RECORD) );
    $this->db->Execute($stmt, $RECORD);
    $pkArray = array("ID" => $RECORD["ID"]);
    $this->findByPk($pkArray, $RECORD);
    print "{success:true, data:".json_encode($RECORD)."}";
    $this->logger->debug("end: ".$this->dcName.".doInsert");
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function doInsert */


public function doUpdate() {
  $this->logger->debug("Start: ".$this->dcName.".doUpdate");
  try {
    $RECORD = array();
    $RECORD["ASSIGNED_TO"] = $this->getRequestParam("ASSIGNED_TO");
    $RECORD["CLOSED"] = $this->getRequestParamBoolean("CLOSED");
    $RECORD["FINISH_DATE"] = $this->getRequestParam("FINISH_DATE");
    $RECORD["ID"] = $this->getRequestParam("ID");
    $RECORD["NOTES"] = $this->getRequestParam("NOTES");
    $RECORD["START_DATE"] = $this->getRequestParam("START_DATE");
    $RECORD["STATUS"] = $this->getRequestParam("STATUS");
    $RECORD["TITLE"] = $this->getRequestParam("TITLE");
    if (empty($RECORD["ID"])) { throw new Exception("Missing value for primary key field ID in DC0019.doUpdate().");}
    $sql = "update TASKS set 
                 CLOSED=:CLOSED
                ,ID=:ID
                ,TITLE=:TITLE
                ,FINISH_DATE=:FINISH_DATE
                ,START_DATE=:START_DATE
                ,STATUS=:STATUS
                ,ASSIGNED_TO=:ASSIGNED_TO
                ,NOTES=:NOTES
    where 
           ID= :ID
    ";
    $stmt = $this->db->prepare($sql);
    $this->logger->debug("update of RECORD: ".$this->logger->map2string($RECORD) );
    $this->db->Execute($stmt, $RECORD);
    $pkArray = array("ID" => $RECORD["ID"]);
    $this->findByPk($pkArray, $RECORD);
    print "{success:true, data:".json_encode($RECORD)."}";
    $this->logger->debug("End: ".$this->dcName.".doUpdate");
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function doUpdate */


public function doDelete() {
  $this->logger->debug("Start: ".$this->dcName.".doDelete");
  try {
    $RECORD["ID"] = $this->getRequestParam("ID");
    if (empty($RECORD["ID"])) { throw new Exception("Missing value for primary key field ID in DC0019.doDelete().");}
    $sql = "delete from TASKS where 
           ID= :ID
    ";
    $stmt = $this->db->prepare($sql);
    $this->db->Execute($stmt, $RECORD);
    print "{success:true, data:".json_encode($RECORD)."}";
    $this->logger->debug("End: ".$this->dcName.".doDelete");
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function doDelete */


public function initNewRecord() {
  try {
    $RECORD["ASSIGNED_TO"] = $this->getRequestParam("ASSIGNED_TO");
    $RECORD["CLOSED"] = $this->getRequestParam("CLOSED");
    $RECORD["CREATEDBY"] = $this->getRequestParam("CREATEDBY");
    $RECORD["CREATEDON"] = $this->getRequestParam("CREATEDON");
    $RECORD["FINISH_DATE"] = $this->getRequestParam("FINISH_DATE");
    $RECORD["ID"] = $this->getRequestParam("ID");
    $RECORD["MODIFIEDBY"] = $this->getRequestParam("MODIFIEDBY");
    $RECORD["MODIFIEDON"] = $this->getRequestParam("MODIFIEDON");
    $RECORD["NOTES"] = $this->getRequestParam("NOTES");
    $RECORD["START_DATE"] = $this->getRequestParam("START_DATE");
    $RECORD["STATUS"] = $this->getRequestParam("STATUS");
    $RECORD["TITLE"] = $this->getRequestParam("TITLE");
    $RECORD["_p_record_status"] = "insert";
    $this->setFieldInitialValues($RECORD, "DC0019");
    print "{success:true, data:".json_encode($RECORD)."}";
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function initNewRecord */

private function findByPk(&$pkCols, &$record) {
    $sql = "select 
                CLOSED
                ,ID
                ,TITLE
                ,FINISH_DATE
                ,START_DATE
                ,STATUS
                ,ASSIGNED_TO
                ,CREATEDON
                ,CREATEDBY
                ,MODIFIEDON
                ,MODIFIEDBY
                ,NOTES
            from TASKS 
         where 
           ID= :ID
            ";
    $rs = $this->db->Execute($sql, $pkCols);
    $row = $rs->FetchRow();
    $this->copyArrayValues($record, $row);
} /* end function findByPk  */

private  $fieldDef = array(
  "CLOSED" => array("DATA_TYPE" => "BOOLEAN")
  ,"ID" => array("DATA_TYPE" => "NUMBER")
  ,"TITLE" => array("DATA_TYPE" => "STRING")
  ,"FINISH_DATE" => array("DATA_TYPE" => "DATE")
  ,"START_DATE" => array("DATA_TYPE" => "DATE")
  ,"STATUS" => array("DATA_TYPE" => "STRING")
  ,"ASSIGNED_TO" => array("DATA_TYPE" => "STRING")
  ,"CREATEDON" => array("DATA_TYPE" => "DATE")
  ,"CREATEDBY" => array("DATA_TYPE" => "STRING")
  ,"MODIFIEDON" => array("DATA_TYPE" => "DATE")
  ,"MODIFIEDBY" => array("DATA_TYPE" => "STRING")
  ,"NOTES" => array("DATA_TYPE" => "STRING")
);


private function readRequest(&$RECORD) {
     if (isset($_REQUEST["ASSIGNED_TO"] )) { $RECORD["ASSIGNED_TO"] = $this->getRequestParam("ASSIGNED_TO"); }
    $RECORD["CLOSED"] = $this->getRequestParamBoolean("CLOSED");
     if (isset($_REQUEST["CREATEDBY"] )) { $RECORD["CREATEDBY"] = $this->getRequestParam("CREATEDBY"); }
     if (isset($_REQUEST["CREATEDON"] )) { $RECORD["CREATEDON"] = $this->getRequestParam("CREATEDON"); }
     if (isset($_REQUEST["FINISH_DATE"] )) { $RECORD["FINISH_DATE"] = $this->getRequestParam("FINISH_DATE"); }
     if (isset($_REQUEST["ID"] )) { $RECORD["ID"] = $this->getRequestParam("ID"); }
     if (isset($_REQUEST["MODIFIEDBY"] )) { $RECORD["MODIFIEDBY"] = $this->getRequestParam("MODIFIEDBY"); }
     if (isset($_REQUEST["MODIFIEDON"] )) { $RECORD["MODIFIEDON"] = $this->getRequestParam("MODIFIEDON"); }
     if (isset($_REQUEST["NOTES"] )) { $RECORD["NOTES"] = $this->getRequestParam("NOTES"); }
     if (isset($_REQUEST["START_DATE"] )) { $RECORD["START_DATE"] = $this->getRequestParam("START_DATE"); }
     if (isset($_REQUEST["STATUS"] )) { $RECORD["STATUS"] = $this->getRequestParam("STATUS"); }
     if (isset($_REQUEST["TITLE"] )) { $RECORD["TITLE"] = $this->getRequestParam("TITLE"); }
} /* end function readRequest  */


public function callProcedure($pName) {
  $this->logger->debug("Start: ".$this->dcName.".callProcedure(".$pName.")");
  try {
    $RECORD = array();
    $this->readRequest($RECORD);
    print "{success:true, data:".json_encode($RECORD)."}";
    $this->logger->debug("End: ".$this->dcName.".callProcedure");
  }catch(Exception  $e) {
      System::sendActionErrorJson( $e->getMessage());
  }
} /* end function callProcedure */

}
?>

File: trunk/ServerJava/src/net/nan21/ebs/dc/DC0019.java
/* N21 eBusiness Suite
 * Copyright: Nan21 Electronics srl
 * Generated content.
 * DC0019 DC Controller: Tasks
 */

package net.nan21.ebs.dc;


import java.util.*;
import javax.servlet.http.HttpServletResponse;
import net.nan21.lib.*;

public class DC0019 extends AbstractDataControl implements IDataControl {

  public void init (HttpRequest request, HttpServletResponse response, HttpSession session, DbManager dbm) throws Exception {
    this._initFields();
    super.init(request, response, session, dbm);
  }

private void preQuery() {
    if (this.request.getParam("QRY_CLOSED") != null && !this.request.getParam("QRY_CLOSED").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("CLOSED like :CLOSED");
      this.queryParams.put("CLOSED",(String)this.request.getParam("QRY_CLOSED"));
    }
    if (this.request.getParam("QRY_ID") != null && !this.request.getParam("QRY_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("ID like :ID");
      this.queryParams.put("ID",(String)this.request.getParam("QRY_ID"));
    }
    if (this.request.getParam("QRY_TITLE") != null && !this.request.getParam("QRY_TITLE").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("TITLE like :TITLE");
      this.queryParams.put("TITLE",(String)this.request.getParam("QRY_TITLE"));
    }
    if (this.request.getParam("QRY_STATUS") != null && !this.request.getParam("QRY_STATUS").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("STATUS like :STATUS");
      this.queryParams.put("STATUS",(String)this.request.getParam("QRY_STATUS"));
    }
}

public void doQuery() throws Exception {
    this.prepareQueryContext();
    this.preQuery();
    this.queryWhere.insert(0, (this.queryWhere.length()>0)?" where ":"");
    String sql = "select "+ 
               " CLOSED"+
               " ,ID"+
               " ,TITLE"+
               " ,FINISH_DATE"+
               " ,START_DATE"+
               " ,STATUS"+
               " ,ASSIGNED_TO"+
               " ,CREATEDON"+
               " ,CREATEDBY"+
               " ,MODIFIEDON"+
               " ,MODIFIEDBY"+
               " ,NOTES"+
           " from TASKS  "+this.queryWhere.toString()+" "+this.queryOrderBy;
    this.writeResultDoQuery(sql);
} 


public void doExport() throws Exception {
    this.prepareQueryContext();
    this.preQuery();
    this.queryWhere.insert(0, (this.queryWhere.length()>0)?" where ":"");
    String sql = "select "+ 
               " TITLE"+
               " ,ASSIGNED_TO"+
               " ,START_DATE"+
               " ,FINISH_DATE"+
               " ,STATUS"+
               " ,CLOSED"+
               " ,NOTES"+
               " ,ID"+
               " ,CREATEDON"+
               " ,CREATEDBY"+
               " ,MODIFIEDON"+
               " ,MODIFIEDBY"+
           " from TASKS  "+this.queryWhere.toString()+" "+this.queryOrderBy;
    this.writeResultDoExport(sql);
}

 	public void fetchRecord()  throws Exception  {
     this.populateRecordPkFromRequest();
     this.findByPk();
       this.writeResultFetchRecord();	 
   }

public void doInsert()  throws Exception {
  this.populateRecordFromRequest(); 
  this.populateRecordWithClientSpecific();
    String sql = "insert into TASKS("+
               "  CLOSED"+
               " ,ID"+
               " ,TITLE"+
               " ,FINISH_DATE"+
               " ,START_DATE"+
               " ,STATUS"+
               " ,ASSIGNED_TO"+
               " ,CREATEDBY"+
               " ,MODIFIEDBY"+
               " ,NOTES"+
           " ) values ( "+
               "  :CLOSED"+
               " ,:ID"+
               " ,:TITLE"+
               " ,:FINISH_DATE"+
               " ,:START_DATE"+
               " ,:STATUS"+
               " ,:ASSIGNED_TO"+
               " ,:CREATEDBY"+
               " ,:MODIFIEDBY"+
               " ,:NOTES"+
    ")";
    this.record.put("ID",   dbm.getSequenceNextValue("seq_task_id")  );
    dbm.executeStatement(sql, this.record);
    this.populateRecordPkFromRecord();
    this.findByPk();
    this.writeResultDoInsert();
}

public void doUpdate() throws Exception {
    this.populateRecordFromRequest();
    this.populateRecordWithClientSpecific();
    String sql = "update TASKS set "+
               "  CLOSED=:CLOSED"+
               " ,ID=:ID"+
               " ,TITLE=:TITLE"+
               " ,FINISH_DATE=:FINISH_DATE"+
               " ,START_DATE=:START_DATE"+
               " ,STATUS=:STATUS"+
               " ,ASSIGNED_TO=:ASSIGNED_TO"+
               " ,NOTES=:NOTES"+
   " where "+
     "      ID= :ID"+
   "";
    dbm.executeStatement(sql, this.record);
    this.populateRecordPkFromRecord();
    this.findByPk();
    this.writeResultDoUpdate();	
}

public void doDelete() throws Exception {
    this.populateRecordPkFromRequest();
   String sql = "delete from TASKS where "+
     "      ID= :ID"+
    "";
    dbm.executeStatement(sql, this.recordPk);
   this.writeResultDoDelete();
}

public void initNewRecord() throws Exception {
    this.populateRecordFromRequest();
    this.record.put("_p_record_status", "insert");
    this.writeResultInitNewRecord();
}

private void findByPk()  throws Exception {
    String sql = "select "+ 
               " CLOSED"+
               " ,ID"+
               " ,TITLE"+
               " ,FINISH_DATE"+
               " ,START_DATE"+
               " ,STATUS"+
               " ,ASSIGNED_TO"+
               " ,CREATEDON"+
               " ,CREATEDBY"+
               " ,MODIFIEDON"+
               " ,MODIFIEDBY"+
               " ,NOTES"+
           " from TASKS "+
        " where "+
     "      ID= :ID"+ 
          "";
    this.record = dbm.executeQuery(sql, this.recordPk).get(0); 
} 


public void doCustomAction(String pName)  throws Exception {
    this.populateRecordFromRequest();
}


	private void  _initFields() {
	  this.fields = new HashMap<String, FieldDef>();
	  this.fields.put("CLOSED", new FieldDef("BOOLEAN"));
	  this.fields.put("ID", new FieldDef("NUMBER"));
	  this.fields.put("TITLE", new FieldDef("STRING"));
	  this.fields.put("FINISH_DATE", new FieldDef("DATE"));
	  this.fields.put("START_DATE", new FieldDef("DATE"));
	  this.fields.put("STATUS", new FieldDef("STRING"));
	  this.fields.put("ASSIGNED_TO", new FieldDef("STRING"));
	  this.fields.put("CREATEDON", new FieldDef("DATE"));
	  this.fields.put("CREATEDBY", new FieldDef("STRING"));
	  this.fields.put("MODIFIEDON", new FieldDef("DATE"));
	  this.fields.put("MODIFIEDBY", new FieldDef("STRING"));
	  this.fields.put("NOTES", new FieldDef("STRING"));
	  String[] _pkFields = {"ID"};
	  this.pkFields = _pkFields;
	  String[] _summaryFields = {};
	  this.summaryFields = _summaryFields;
	  this.queryResultSize = 20;
	}

}
